File: 2018-11-15/IBMGUY/longest_substring/longest_substring.swift

func lengthOfLongestSubstring(s: String) -> Int {
    if s.count < 2 {
        return s.count
    }
    // 🐇 & 🐢 represent indices in the string that run left to right
    var 🐢 = 0
    var 🐇 = 1
    // used to detect when 🐇 approaches a duplicate letter
    var lettersBeingTracked = Set<Character>() // CharacterSet is nice but not appropriate here
    // initialize longestSubstring with the first two letters in the string
    var longestSubstring = getSubstringFromIndices(s: s, start: 🐢, end: 🐇)
    // unless they are the same character
    if s[🐢] == s[🐇] {
        longestSubstring = Substring(String((s[🐇]))) // casting Character -> String -> Substring
    }

    while 🐇 < s.count {
        🐇 += 1
        var nextLetter: Character?
        if 🐇 == s.count {
            nextLetter = s[🐇 - 1]
        } else {
            nextLetter = s[🐇]
        }

        if lettersBeingTracked.contains(nextLetter!) {
            // found a duplicate, so move the turtle up past the last time the duplicate was seen
            while 🐢 < 🐇 {
                🐢 += 1
                if s[🐢 - 1] == nextLetter {
                    break
                }
                // letters the 🐢 walks past are no longer being tracked in the dynamic substring
                lettersBeingTracked.remove(s[🐢 - 1])
            }
        } else {
            // a new substring is encountered, so check it against the longest
            let currentSubstring: Substring = getSubstringFromIndices(s: s, start: 🐢, end: 🐇)
            if currentSubstring.count > longestSubstring.count {
                longestSubstring = currentSubstring
            }

            lettersBeingTracked.insert(nextLetter!)
        }
    }

    return longestSubstring.count
}

func getSubstringFromIndices(s: String, start: Int, end: Int) -> Substring {
    let startIndex = s.index(s.startIndex, offsetBy: start)
    let endIndex = s.index(s.endIndex, offsetBy: end - s.count - 1)
    return s[startIndex..<endIndex]
}

extension String {
    subscript (i: Int) -> Character {
        return self[index(startIndex, offsetBy: i)]
    }
}

// this will take at most 2 iterations on the characters of string s
// when a duplicate is found, the newest valid substring starts after the first instance of the duped letter, not the latest letter
// edge inputs: "au", "aab", "", extremely long inputs

File: 2018-11-15/IBMGUY/longest_substring/longest_substring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):

        turtle_idx = 0
        hare_idx = 1

        # don't need to check inputs less than 2
        if len(s) < 2:
            return len(s)

        letters_being_tracked = set([s[0], s[1]])

        # default longest substring starts as length 2 unless the first two chars are the same
        if s[turtle_idx] == s[hare_idx]:
            longest_substring = s[hare_idx]
            turtle_idx += 1
        else:
            longest_substring = s[turtle_idx:hare_idx+1]

        while hare_idx < len(s):
            hare_idx += 1
            # check if the next letter will be a duplicate
            if hare_idx == len(s):
                next_letter = s[-1]
            else:
                next_letter = s[hare_idx]


            if next_letter in letters_being_tracked:
                # move the turtle up to the last time this repeating letter occurred
                while turtle_idx < hare_idx:
                    turtle_idx += 1
                    if s[turtle_idx - 1] == next_letter:
                        break

                    # remove previously seen letters after the turtle 'passes' them
                    letters_being_tracked.remove(s[turtle_idx - 1])
            else:
                # check substring against longest
                current_substring = s[turtle_idx:hare_idx + 1]

                if len(current_substring) > len(longest_substring):
                    longest_substring = current_substring

                letters_being_tracked.add(s[hare_idx])


        return len(longest_substring)


            
